src/main/java/org/opentripplanner/graph_builder/impl/shapefile/NihShapefileStreetGraphBuilderImpl.java

```diff
@@ -0,0 +1,457 @@
+                    foundStart = true;
+                    startEdges.add(edge);
+                    farStartVertices.add(to);
+                } else if (toLabel.equals(startIntersection.getLabel())) {
+                    foundStart = true;
+                    startEdges.add(edge);
+                    farStartVertices.add(from);
+                }
+
+                if (fromLabel.equals(endIntersection.getLabel())) {
+                    foundEnd = true;
+                    endEdges.add(edge);
+                    farEndVertices.add(to);
+                } else if (toLabel.equals(endIntersection.getLabel())) {
+                    foundEnd = true;
+                    endEdges.add(edge);
+                    farEndVertices.add(from);
+                }
+            }
+
+            if (!foundStart || !foundEnd) {
+                LOG.warn("Could not find start or end for {}: {}, segment {}", wayLabel, wayName, segmentId);
+                continue;
+            }
+
+            // check if the start and end intersections are connected by a single graph segment
+            // (if so, there will be two edges found, one the reverse of the other)
+            boolean foundMatch = false;
+            for (StreetEdge start : startEdges) {
+                for (StreetEdge end : endEdges) {
+                    if (start.isEquivalentTo(end)) {
+                        foundMatch = true;
+                        matchedEdges.add(start);
+                    }
+                }
+            }
+
+            if (!foundMatch) {
+                // check if the start and end intersections are connected by two edges (that share a vertex in between)
+                // note the shared vertex, if found, in order to add edges that go to/from it later
+                String sharedVertex = null;
+                for (Vertex start : farStartVertices) {
+                    for (Vertex end : farEndVertices) {
+                        if (start.getLabel().equals(end.getLabel())) {
+                            sharedVertex = start.getLabel();
+                            break;
+                        }
+                    }
+                }
+
+                if (sharedVertex != null) {
+                    LOG.info("Start and end vertex share a midpoint at {}", sharedVertex);
+                    // go grab only edges that go between shared vertex and start or end edge
+                    List<StreetEdge> candidates = new ArrayList<>();
+                    candidates.addAll(startEdges);
+                    candidates.addAll(endEdges);
+                    for (StreetEdge edge: candidates) {
+                        if (edge.getFromVertex().getLabel().equals(sharedVertex) ||
+                                edge.getToVertex().getLabel().equals(sharedVertex)) {
+
+                            matchedEdges.add(edge);
+                        }
+                    }
+                } else {
```

Comment:
Yes, calculating SPT is slower, and not needed most of the time.  I'm more concerned about the SPT search not being the most reliable method; what if there happens to be a more direct path that doesn't happen to travel the way we're looking at?  Then it would be necessary to force it to calculate more paths, then go walk them.